File: USACO Practice/src/connectingtwobarns/connectingtwobarns.java
//package connectingtwobarns;

import java.util.*;
import java.io.*;

public class connectingtwobarns {
	
	static StreamTokenizer in;
	static int nextInt() throws IOException {
		in.nextToken();
		return (int) in.nval;
	}

	public static void main(String[] args) throws IOException {
		
		//USACO 2021 December Contest Silver
		
		//graphs. idk why this one is timing out
		
		//BufferedReader fin = new BufferedReader(new InputStreamReader(System.in));
		in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
		int t = nextInt();
		StringBuilder fout = new StringBuilder();
		while(t-- > 0) {
			int n = nextInt();
			int m = nextInt();
			ArrayList<ArrayList<Integer>> c = new ArrayList<ArrayList<Integer>>();
			for(int i = 0; i < n; i++) {
				c.add(new ArrayList<Integer>());
			}
			for(int i = 0; i < m; i++) {
				//st = new StringTokenizer(fin.readLine());
				int a = nextInt() - 1;
				int b = nextInt() - 1;
				c.get(a).add(b);
				c.get(b).add(a);
			}
			ArrayList<ArrayList<Integer>> groups = new ArrayList<ArrayList<Integer>>();
			boolean[] v = new boolean[n];
			for(int i = 0; i < n; i++) {
				if(!v[i]) {
					ArrayList<Integer> curGroup = new ArrayList<Integer>();
					Stack<Integer> s = new Stack<Integer>();
					s.add(i);
					v[i] = true;
					while(s.size() != 0) {
						int cur = s.pop();
						curGroup.add(cur);
						for(int a : c.get(cur)) {
							if(!v[a]) {
								v[a] = true;
								s.add(a);
							}
						}
					}
					groups.add(curGroup);
				}
			}
			
			int hasStart = 0;
			int hasEnd = 0;
			for(int i = 0; i < groups.size(); i++) {
				for(int j : groups.get(i)) {
					if(j == 0) {
						hasStart = i;
					}
					if(j == n - 1) {
						hasEnd = i;
					}
				}
			}
			if(hasStart == hasEnd) {
				fout.append("0\n");
				continue;
			}
			for(int i = 0; i < groups.size(); i++) {
				groups.get(i).sort((a, b) -> Integer.compare(a, b));
			}
			//System.out.println(groups);
			long[] fromStart = new long[groups.size()];
			long[] toEnd = new long[groups.size()];
			for(int i = 0; i < groups.size(); i++) {
				int a = 0;
				int b = 0;
				int minDiff = Integer.MAX_VALUE;
				while(true) {
					//System.out.println(a + " " + b);
					int aVal = groups.get(hasStart).get(a);
					int bVal = groups.get(i).get(b);
					//System.out.println(aVal + " " + bVal);
					minDiff = Math.min(minDiff, Math.abs(aVal - bVal));
					if(minDiff == 0 || minDiff == 1) {
						break;
					}
					if(aVal < bVal) {
						a++;
					}
					else {
						b++;
					}
					if(a > groups.get(hasStart).size() || b > groups.get(i).size()) {
						break;
					}
				}
				fromStart[i] = minDiff;
			}
			for(int i = 0; i < groups.size(); i++) {
				int a = 0;
				int b = 0;
				int minDiff = Integer.MAX_VALUE;
				while(true) {
					int aVal = groups.get(hasEnd).get(a);
					int bVal = groups.get(i).get(b);
					minDiff = Math.min(minDiff, Math.abs(aVal - bVal));
					if(minDiff == 0 || minDiff == 1) {
						break;
					}
					if(aVal < bVal) {
						a++;
					}
					else {
						b++;
					}
					if(a > groups.get(hasEnd).size() || b > groups.get(i).size()) {
						break;
					}
				}
				toEnd[i] = minDiff;
			}
			long ans = Long.MAX_VALUE;
			for(int i = 0; i < groups.size(); i++) {
				ans = (long) Math.min(ans, Math.pow(fromStart[i], 2l) + Math.pow(toEnd[i], 2l));
			}
			fout.append(ans).append("\n");
			
			
		}
		System.out.print(fout);
	}
}

File: - cpp -/connectingtwobarns.cpp

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <bits/stdc++.h>

typedef long long ll;

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t;
    cin >> t;
    while(t-- > 0){
        int n, m;
        cin >> n >> m;
        vector<vector<int>> c = vector<vector<int>>(n);
        for(int i = 0; i < n; i++){
            c.push_back(vector<int>());
        }
        for(int i = 0; i < m; i++){
            int a, b;
            cin >> a >> b;
            c[a - 1].push_back(b - 1);
            c[b - 1].push_back(a - 1);
        }
        vector<vector<int>> groups = vector<vector<int>>();
        vector<bool> v = vector<bool>(n);
        for(int i = 0; i < n; i++){
            if(!v[i]){
                groups.push_back(vector<int>());
                stack<int> s = stack<int>();
                s.push(i);
                v[i] = true;
                while(s.size() != 0){
                    int cur = s.top();
                    s.pop();
                    groups[groups.size() - 1].push_back(cur);
                    for(int a : c[cur]){
                        if(!v[a]){
                            v[a] = true;
                            s.push(a);
                        }
                    }
                }
            }
        }
        int hasStart, hasEnd;
        for(int i = 0; i < groups.size(); i++){
            for(int j : groups[i]){
                if(j == 0){
                    hasStart = i;
                }
                if(j == n - 1){
                    hasEnd = i;
                }
            }
        }
        if(hasStart == hasEnd){
            cout << "0\n";
            continue;
        }
        for(int i = 0; i < groups.size(); i++){
            sort(groups[i].begin(), groups[i].end());
        }
        vector<ll> fromStart = vector<ll>(groups.size());
        vector<ll> toEnd = vector<ll>(groups.size());
        for(int i = 0; i < groups.size(); i++){
            int a = 0;
            int b = 0;
            ll minDiff = 1000000;
            while(true){
                int aVal = groups[hasStart][a];
                int bVal = groups[i][b];
                minDiff = min(minDiff, (ll)abs(aVal - bVal));
                if(minDiff == 0){
                    break;
                }
                if(aVal < bVal){
                    if(a + 1 < groups[hasStart].size()){
                        a ++;
                    }
                    else {
                        break;
                    }
                }
                if(bVal < aVal){
                    if(b + 1 < groups[i].size()){
                        b ++;
                    }
                    else {
                        break;
                    }
                }
            }
            fromStart[i] = (ll) minDiff * (ll) minDiff;
        }
        for(int i = 0; i < groups.size(); i++){
            int a = 0;
            int b = 0;
            ll minDiff = 1000000;
            while(true){
                int aVal = groups[hasEnd][a];
                int bVal = groups[i][b];
                minDiff = min(minDiff, (ll)abs(aVal - bVal));
                if(minDiff == 0){
                    break;
                }
                if(aVal < bVal){
                    if(a + 1 < groups[hasEnd].size()){
                        a ++;
                    }
                    else {
                        break;
                    }
                }
                if(bVal < aVal){
                    if(b + 1 < groups[i].size()){
                        b ++;
                    }
                    else {
                        break;
                    }
                }
            }
            toEnd[i] = (ll) minDiff * (ll) minDiff;
        }
        ll ans = fromStart[0] + toEnd[0];
        for(int i = 1; i < groups.size(); i++){
            ans = min(ans, fromStart[i] + toEnd[i]);
        }
        cout << ans << endl;
    }    
    
    return 0;
}
